metrics/bigquery.py

```diff
@@ -50,67 +51,171 @@ def validate_metric_name(name):
+        # (For backfilling)
+        measure = config.get('measurements', {}).get('backfill')
+        if not measure:
+            return
+        if self.influx:
+            # To get the last data point timestamp we must also fetch a field.
+            # So first find a field that we can query if the metric exists.
+            points = self.influx.query('show field keys from %s limit 1' % measure)
+            points = list(points.get_points())
+
+            field = points and points[0].get('fieldKey')
+            last_time = None
+            if field:
+                results = self.influx.query(
+                    'select last(%s), time from %s limit 1' % (field, measure)
                 )
-            influx.push(points, "metrics")
+                last_time = next(results.get_points(), {}).get('time')
+                if last_time:
+                    # format time properly
+                    last_time = time.strptime(last_time, '%Y-%m-%dT%H:%M:%SZ')
+                    last_time = calendar.timegm(last_time)
+            if not last_time:
+                last_time = int(time.time() - (60*60*24*self.backfill_days))
         else:
-            print >>sys.stderr, 'Skipping upload of %s, no influxdb config.\n' % metric
-
-    def copy(src, dst):
-        """Use gsutil to copy src to test with minimal caching."""
-        check([
-            'gsutil', '-h', 'Cache-Control:max-age=60', 'cp', src, dst])
+            # InfluxDB is not enabled so skip backfill, just query for last day.
+            last_time = int(time.time() - (60*60*24))
 
-    copy(raw, os.path.join(prefix, metric, raw))
-    copy(filtered, os.path.join(prefix, metric, filtered))
-    copy(filtered, os.path.join(prefix, latest))
+        # replace tag with formatted time
+        config['query'] = config['query'].replace('<LAST_DATA_TIME>', str(last_time))
 
 def all_configs(search='**.yaml'):
     """Returns config files in the metrics dir."""
     return glob.glob(os.path.join(
         os.path.dirname(__file__), 'configs', search))
 
-def main(configs, project, bucket_path):
-    """Loads metric config files and runs each metric."""
-    if not project:
-        raise ValueError('project', project)
-    if not bucket_path:
-        raise ValueError('bucket_path', bucket_path)
+def make_influx_client():
+    """Make an InfluxDB client from config at path $VELODROME_INFLUXDB_CONFIG"""
+    if 'VELODROME_INFLUXDB_CONFIG' not in os.environ:
+        return None
 
-    if 'VELODROME_INFLUXDB_CONFIG' in os.environ:
-        influx = influxdb.Pusher.from_config(os.environ['VELODROME_INFLUXDB_CONFIG'])
-    else:
-        influx = None
+    with open(os.environ['VELODROME_INFLUXDB_CONFIG']) as config_file:
+        config = json.load(config_file)
+
+    def check_config(field):
+        if not field in config:
+            raise ValueError('DB client config needs field \'%s\'' % field)
+    check_config('host')
```

Comment:
btw I like this pattern because it shows a better strack trace! (over using a for loop)